src/NUnitEngine/nunit-agent/Program.cs

```diff
@@ -130,58 +122,65 @@ public static void Main(string[] args)
             log.Info("Initializing Services");
             engine.Initialize();
 
-            log.Info("Starting RemoteTestAgent");
-            Agent = new RemoteTestAgent(AgentId, AgencyUrl, engine.Services);
-
+            log.Info("Starting AgentServer");
             try
             {
-                if (Agent.Start())
-                    WaitForStop();
-                else
+                using (var server = AgentServer.Start(IPAddress.Loopback, engine.Services.GetService<ITestRunnerFactory>()))
                 {
-                    log.Error("Failed to start RemoteTestAgent");
-                    Environment.Exit(AgentExitCodes.FAILED_TO_START_REMOTE_AGENT);
+                    // Write to the console to let the parent process know which port Windows assigned
+                    Console.WriteLine("Listening on " + server.ListeningOn);
+                    log.Info("Listening on " + server.ListeningOn);
```

Comment:
I'm a little confused as to why writing to stdout is necessary, sorry. Can the engine not just bind to the port first, and pass the port number in on the agent command line args?

Reply:
I agree with @ChrisMaddock.  It would be better to not write to stdout.

Reply:
We'll get to the point hopefully very soon where the user is starting the TCP server first and passing the port number as command-line argument to the console runner. I'd like to keep them decoupled in this way.
Let's say we did that though. If we had a separate mode where the console runner bound to a port, we'd have to rely on the bound socket getting inherited into the child agent process. We'd have to duplicate the socket handle as inheritable and close the non-inheritable one. It felt much cleaner to me to have the console runner communicate with the agent just like a user would.

What is it about writing to stdout that could be problematic? I particularly liked this part and thought it would even be a selling point 😛, but of course I'm willing to do whatever.

Reply:
I'm confused by this as well. The existing remoting facility runs on top of tcp which US set up without other communication channel. The server has a single endpoint where clients register and are then given a custom connection on which they receive commands and return results. All that is done by remoting. It's just that we need to do it in our own code now.

Reply:
Many protocols do the same thing.  FTP, for instance.  They have a pre-defined port for initiating communications, and then negotiate a port to use for the commands and data.

We could do the same, and then allow a command line override for that port if really necessary.

Reply:
That is still seriously complex compared to what it could be.

- Runner spins up a TCP server and passes the port it gets to the child agent process when it starts
- Agent connects back to the runner using TCP
- Agent spins up a TCP server and passes the port it gets to the parent process via the existing connection
- Runner and agent both disconnect from the original connection and begin talking on the second one

None of this seems necessary to me. In this scenario, the runner is not behaving like the thin client that I envision it to be. The runner only lasts as long as the test run it's executing. Once we start having agents that live longer, it will make less sense for all that to be part of the protocol.

It would be cool if we could keep it simplified so that we always have a client mindset when talking about the console runner and a server mindset when talking about the agent.
Speaking of what is confusing, this will best model people's expectations of such an agent, too. Other build and test agents seem to typically be structured this way.

Reply:
I'm picturing the responsibilities as linear, fitting both scenarios without complicating the inner part (in bold).
The first three points are analogous between the two scenarios, and the protocol responsibility is separated in such a way that the final point (our custom protocol) can be entirely independent of the scenario.

Scenario A: User wants tests to run tests in an existing agent 

- The user installs the agent as a service on a dedicated test server,
  or starts an Android emulator and opens an NUnit app hosting an agent,
  or starts an instance of VS with an agent extension to do UI tests in the VS process.

- In each of these cases, the user knows the IP address and port number due to either choosing the port number when starting/scheduling the agent, or seeing the port number and IP address when the Xamarin app starts, etc.

- The user passes the IP address and port number (or just port number if localhost) to the console runner.

- **The runner acts as a fairly thin client.**

Scenario B: User wants the runner to start and stop an agent

- The user starts the console runner without specifying an existing agent.

- The console runner acts on the user's behalf by managing the lifetime of the agent process.

- The console runner knows the IP address of the agent process because the runner process started it.

- **The runner acts as a fairly thin client.**

Reply:
I take it that by agent, you are referring only to the existing nunit-agent executable used to run in a separate process. I was thinking more generally, in the sense that all execution is performed by an agent, but I'll stick with the restricted definition for now so we're on the same page.

In the relationship between the TestAgency and agents, it's the  agency which is the server, even though the agents are seen as "doing something" for the runner. This apparent reversal is fairly common in designs where the process to perform a task is actually spawned by the process that needs something done.

This isn't just a distinction of terms because it's the server's job, not the client's, to provide communication endpoints. The TestAgency "serves" multiple clients while the agent   only exists for short time to deal with a single runner.

We had always envisioned having other kinds of agents, on separate machines, functioning as actual servers, but we aren't doing that here - or if we are, then I'd say too much for one PR.

WRT encapsulation rather than replacement, I agree that the latter is simpler. But it leaves unsolved a fundamental problem in the engine, which is that it can only communicate in one way.

Reply:
It was my understanding that was the goal, to allow the agent to run on separate machines functioning as an actual server.

That is what I need for CF, and what I think @jnm2 meant when he discussed a Xamarian emulator and an Android device.

Reply:
@jnm2 Is that what you are trying to do? At one time we had a server project, which has been removed from the current repo. Back then, we used "server" to mean what @oznetmaster is talking about and "agent" for the current single-use, TestAgency-initiated processes.

I hadn't seen any indication that you were hoping to have one "agent" that works both ways, but since @oznetmaster 's comment seems to suggest that, I'll throw in my two cents, which is that you would be trying to do too much in one bit of code. Two such different "agents" would obviously share a lot of code - in fact, the code that I was calling "agent" or "core" in a separate spike I did. But they would manage their communications entirely differently. IMO, you'll have better result by limiting the PR to one thing.

Reply:
I did think that the console would open the port and listen, then pass the port on the command line to the agent. I can see some of your points on wanting to do it the other way around, but it does introduce some difficulty. What if the agent is on a different machine or device?

Reply:
A bit of history, in case it helps...

We started out with two concepts... agents and servers...

Agents were to be those local processes initiated by the TestAgency, a server itself, which called back to the test agency at an endpoint contained in their argument list, received a job to run, ran it to completion, and exited when the tests were finally unloaded.

Servers, for which we had a separate project, were intended to be long-running and findable by the desktop runner. They were aimed at running on a separate machine although they could obviously be tested on the same machine. The desktop runner and engine would function as a client to such servers. They would clean up afterward and then accept new jobs.

I actually tried to do what @jnm2 is doing first. Having the short-term agents function as servers required extra communications. I saw some code written by @jcansdale for his original open source extension, where the roles were reversed, with the assembly running the tests functioning as a client to a little server embedded in the extension. I took that idea and never looked back. (Thanks @jcansdale ) 

For a long time we maintained code for what was basically a null server. It ran, you could contact it but it did nothing. We eventually dropped that project - I'm not sure if we did it for NUnit 3 or prior to that time - because nobody was working on int.

IMO, the long-running server and the short-running agent are two different things. The former is obviously more complicated. I think we need both. Most of the code will be the same, just combined in different ways and the separation will make the more complex server process just a little less complex.

BTW, this has nothing to do with whatever architecture current CI systems use for "agents". This goes back to something like 2004 and the terms were pure inventions at the time.